onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/softmax_op_builder.cc

```diff
@@ -119,20 +171,42 @@ bool SoftMaxOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& /* initiali
   if (!GetShape(node_unit.Inputs()[0].node_arg, input_shape))
     return false;
 
-  const auto input_size = input_shape.size();
-  if (input_size != 2 && input_size != 4) {
-    LOGS_DEFAULT(VERBOSE) << "SoftMax only support 2d/4d shape, input is "
-                          << input_size << "d shape";
-    return false;
-  }
+  if (node_unit.SinceVersion() < 13) {
+    // if opset < 13 the ONNX spec coerces to 2D based on axis, so we will manually do that when adding to the model
+    // and use axis of 1. A 2D input with axis==1 works with all NNAPI versions.
+  } else {
+    const auto input_size = input_shape.size();
+
+    if (input_size > 4) {
+      LOGS_DEFAULT(VERBOSE) << "Softmax only supports maximum 4d input. input is " << input_size << "d";
+      return false;
+    }
 
-  if (params.android_feature_level < ANEURALNETWORKS_FEATURE_LEVEL_3) {
     NodeAttrHelper helper(node_unit);
     int32_t axis = helper.Get("axis", 1);
-    if (axis != 1) {
-      LOGS_DEFAULT(VERBOSE)
-          << "SoftMax only support axis 1 on Android API level: " << params.android_feature_level
-          << " input axis: " << axis;
+    if (axis < 0) {
+      axis = static_cast<int32_t>(HandleNegativeAxis(axis, input_shape.size()));
+    }
+
+    if (params.android_feature_level < ANEURALNETWORKS_FEATURE_LEVEL_3) {
+      if (input_size != 2 && input_size != 4) {
+        LOGS_DEFAULT(VERBOSE) << "Softmax only support 2d/4d shape, input has " << input_size << "d shape";
+        return false;
+      }
+
+      if (axis != 1) {
+        LOGS_DEFAULT(VERBOSE) << "Softmax only support axis 1 on Android API level: " << params.android_feature_level
+                              << " input axis: " << axis;
+        return false;
+      }
+    }
+
+    if (static_cast<uint32_t>(axis) != input_size - 1) {
```

Comment:
are other axis values supported starting from ANEURALNETWORKS_FEATURE_LEVEL_3?

Reply:
https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0a2bfbb83a537701e2843a3d5004250c2c

> 2: An optional [ANEURALNETWORKS_INT32](https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaf06d1affd33f3bc698d0c04eceb23298acb165f692e9d189c820ceba84b3c21dc) scalar, default to -1, specifying the dimension the activation would be performed on. Negative index is used to specify axis from the end (e.g. -1 for the last axis). Must be in the range [-n, n). Available since NNAPI feature level 3.

So you can specify any axis, but due to the ONNX spec change the current implementation only supports the last dimension being the axis.

Reply:
maybe I'm missing something - I don't understand the difference between ANEURALNETWORKS_SOFTMAX's input 2 vs. ONNX SoftMax-13's axis attribute.

https://github.com/onnx/onnx/blob/main/docs/Operators.md#Softmax

> The "axis" attribute indicates the dimension along which Softmax will be performed.

Reply:
Updated as per discussion. NNAPI Softmax is equivalent to ONNX Softmax-13 in that it performs the Softmax on the specific axis. 

For older NNAPI versions a 2D input was internally converted to 4D by inserting two dims of size 1 in the middle so that the second dim becomes the last dim, and the default axis of -1 works. i.e. NNAPI wasn't coalescing 4D to 2D, but rather unsqueezing 2D to 4D. 

Also confusion from the NNAPI spec/code where the term 'batch' equates to all dimensions up to the axis rather than the typical usage of the batch being only the first dimension.